File: app/python/src/utils/transfer_client.py
#!/usr/bin/env python3
# laurent.martin.aspera@fr.ibm.com
# Common library for sample scripts
# Helper methods to get API environment according to config file
# Simplified function to start transfer and wait for it to finish
import os
import re
import json
import time
import grpc
import logging
import subprocess
import utils.configuration
from urllib.parse import urlparse
import warnings
warnings.filterwarnings("ignore", ".*obsolete", UserWarning, "google.protobuf.runtime_version")

# before stub import: protobuf: avoid incompatibility of version, use pure python implementation
os.environ['PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION'] = 'python'
# avoid message: 'Other threads are currently calling into gRPC, skipping fork() handlers'
os.environ['GRPC_ENABLE_FORK_SUPPORT'] = 'false'

# import gRPC stubs (Transfer SDK API), make sure it is in PYTHONPATH
import transferd_pb2_grpc as transfer_manager_grpc  # noqa: E4
import transferd_pb2 as transfer_manager  # noqa: E4

ASCP_LOG_FILE = "aspera-scp-transfer.log"
DEBUG_HTTP = False


class TransferClient:
    '''Transfer Client using Aspera Transfer SDK'''

    def __init__(self, config):
        self._config = config
        sdk_url = urlparse(self._config.param('trsdk', 'url'))
        self._server_address = sdk_url.hostname
        self._server_port = sdk_url.port
        self._transfer_daemon_process = None
        self._transfer_service = None
        self._daemon_name = os.path.basename(self._config.get_path('sdk_daemon'))
        self._daemon_log = os.path.join(self._config._log_folder, f"{self._daemon_name}.log")

    def create_config_file(self, conf_file):
        '''
        see https://developer.ibm.com/apis/catalog/aspera--aspera-transfer-sdk/Configuration%20File
        '''
        config_info = {
            'address': self._server_address,
            'port': self._server_port,
            'log_directory': self._config._log_folder,
            'log_level': self._config.param('trsdk', 'level'),
            'fasp_runtime': {
                'use_embedded': True,
                'log': {
                    'dir': self._config._log_folder,
                    'level': ascp_level(self._config.param('trsdk', 'ascp_level')),
                },
            },
        }
        config_data = json.dumps(config_info)
        logging.debug('config: %s', config_data)
        with open(conf_file, 'w') as the_file:
            the_file.write(config_data)

    def start_daemon(self):
        '''
        Start transfer manager daemon if not already running

        @return gRPC client
        '''
        file_base = os.path.join(self._config._log_folder, self._daemon_name)
        conf_file = f'{file_base}.conf'
        out_file = f'{file_base}.out'
        err_file = f'{file_base}.err'
        command = ' '.join([
            self._config.get_path('sdk_daemon'),
            '--config',
            conf_file,
        ])
        logging.debug('daemon out: %s', out_file)
        logging.debug('daemon err: %s', err_file)
        logging.debug('daemon log: %s', self._daemon_log)
        logging.debug('ascp log: %s', os.path.join(
            self._config._log_folder, ASCP_LOG_FILE))
        logging.debug('command: %s', command)
        self.create_config_file(conf_file)
        logging.info('Starting daemon...')
        self._transfer_daemon_process = subprocess.Popen(
            command,
            shell=True,
            stdout=open(out_file, 'w'),
            stderr=open(err_file, 'w'),
        )
        # give time for startup
        time.sleep(2)
        exit_status = self._transfer_daemon_process.poll()
        if exit_status is not None:
            logging.error('Daemon not started.')
            logging.error('Exited with code: %s', exit_status)
            logging.error('Check daemon log: %s', self._daemon_log)
            logging.error(utils.configuration.last_file_line(self._daemon_log))
            raise Exception('daemon startup failed')
        logging.info('Daemon started: %s', self._transfer_daemon_process.pid)
        # port zero means: listen on any available port, but we need to know the real port
        if self._server_port == 0:
            last_line = utils.configuration.last_file_line(self._daemon_log)
            log_info = json.loads(last_line)
            port_match = re.search(r":(\d+)", log_info["msg"])
            if not port_match:
                raise Exception('Could not read listening port from log file')
            self._server_port = port_match.group(1)
            logging.info('Allocated server port: %s', self._server_port)

    def connect_to_daemon(self):
        '''Connect to transfer manager daemon'''
        channel_address = f'{self._server_address}:{self._server_port}'
        logging.info('Connecting to %s on: %s ...', self._daemon_name, channel_address)
        # create a connection to the transfer manager daemon
        channel = grpc.insecure_channel(channel_address)
        try:
            grpc.channel_ready_future(channel).result(timeout=5)
        except grpc.FutureTimeoutError:
            logging.error('Failed to connect')
            raise Exception('failed to connect.')
        # channel is ok, let's get the stub
        self._transfer_service = transfer_manager_grpc.TransferServiceStub(channel)
        logging.info('Connected !')

    def startup(self):
        '''Start and connect to transfer manager daemon'''
        if self._transfer_service is None:
            self.start_daemon()
            self.connect_to_daemon()
        return self

    def shutdown(self):
        '''Shutdown transfer manager daemon, if needed'''
        if self._transfer_service is None:
            self._transfer_service = None
        if self._transfer_daemon_process is not None:
            logging.info('Shutting down daemon...')
            # self._transfer_daemon_process.send_signal(signal.CTRL_C_EVENT)
            # self._transfer_daemon_process.terminate()
            self._transfer_daemon_process.kill()
            self._transfer_daemon_process.wait()
            self._transfer_daemon_process = None

    def start_transfer(self, transfer_spec):
        '''Start a transfer and return transfer id'''
        ts_json = json.dumps(transfer_spec)
        logging.debug('ts: %s', ts_json)
        # create a transfer request
        transfer_request = transfer_manager.TransferRequest(
            transferType=transfer_manager.FILE_REGULAR,
            config=transfer_manager.TransferConfig(),
            transferSpec=ts_json,
        )
        # send start transfer request to transfer manager daemon
        transfer_response = self._transfer_service.StartTransfer(transfer_request)
        self.throw_on_error(transfer_response.status, transfer_response.error)
        return transfer_response.transferId

    def wait_transfer(self, transfer_id):
        '''Wait for transfer completion'''
        logging.debug('transfer started with id %s', transfer_id)
        # monitor transfer status
        for transfer_info in self._transfer_service.MonitorTransfers(
                transfer_manager.RegistrationRequest(
                    filters=[transfer_manager.RegistrationFilter(
                        transferId=[transfer_id])]
                )):
            # logging.debug('transfer info %s', transfer_info)
            # check transfer status in response, and exit if it's done
            status = transfer_info.status
            logging.info('transfer: %s', transfer_manager.TransferStatus.Name(status))
            self.throw_on_error(status, transfer_info.error)
            if status == transfer_manager.COMPLETED:
                break

    def start_transfer_and_wait(self, t_spec):
        '''One-call simplified procedure to start daemon, transfer and wait for it to finish'''
        # TODO: remove when transfer sdk bug fixed
        # t_spec['http_fallback'] = False
        self.startup()
        self.wait_transfer(self.start_transfer(t_spec))

    def throw_on_error(self, status, error):
        '''raise exception if status contains an error'''
        if status == transfer_manager.TransferStatus.FAILED:
            logging.error(utils.configuration.last_file_line(self._daemon_log))
            raise Exception("transfer failed: " + error.description)
        if status == transfer_manager.TransferStatus.UNKNOWN_STATUS:
            raise Exception("unknown transfer id: " + error.description)


def ascp_level(level_string):
    if level_string == 'info':
        return 0
    elif level_string == 'debug':
        return 1
    elif level_string == 'trace':
        return 2
    else:
        raise Exception('Invalid ascp_level: ' + level_string)

File: app/ruby/src/utils/transfer_client.rb
# frozen_string_literal: true

require 'json'
require 'logger'
require 'fileutils'
require 'securerandom'
require 'open3'
require 'grpc'
require 'uri'
require_relative '../transferd_services_pb'

module Utils
  class TransferClient
    ASCP_LOG_FILE = 'aspera-scp-transfer.log'
    DEBUG_HTTP = false
    def initialize(config)
      @config = config
      sdk_url = URI.parse(@config.param('trsdk', 'url'))
      @server_address = sdk_url.host
      @server_port = sdk_url.port
      @transfer_daemon_process = nil
      @transfer_service = nil
      @daemon_name = File.basename(@config.get_path('sdk_daemon'))
      @daemon_log = File.join(@config.log_folder, "#{@daemon_name}.log")
      @logger = config.logger
    end

    def create_config_file(conf_file)
      config_info = {
        'address' => @server_address,
        'port' => @server_port,
        'log_directory' => @config.log_folder,
        'log_level' => @config.param('trsdk', 'level'),
        'fasp_runtime' => {
          'use_embedded' => true,
          'log' => {
            'dir' => @config.log_folder,
            'level' => ascp_level(@config.param('trsdk', 'ascp_level'))
          }
        }
      }
      File.write(conf_file, JSON.pretty_generate(config_info))
    end

    def start_daemon
      file_base = File.join(@config.log_folder, @daemon_name)
      conf_file = "#{file_base}.conf"
      out_file  = "#{file_base}.out"
      err_file  = "#{file_base}.err"

      command = [
        @config.get_path('sdk_daemon'),
        '--config', conf_file
      ]

      @logger.debug("daemon out: #{out_file}")
      @logger.debug("daemon err: #{err_file}")
      @logger.debug("daemon log: #{@daemon_log}")
      @logger.debug("ascp log: #{File.join(@config.log_folder, ASCP_LOG_FILE)}")
      @logger.debug("command: #{command.join(' ')}")

      create_config_file(conf_file)
      @logger.info('Starting daemon...')

      @transfer_daemon_process = Process.spawn(*command,
                                               out: out_file,
                                               err: err_file)
      sleep 2

      _, status = Process.wait2(@transfer_daemon_process, Process::WNOHANG)
      if status
        @logger.error("Daemon not started, exit code=#{status.exitstatus}")
        @logger.error("Check daemon log: #{@daemon_log}")
        raise 'daemon startup failed'
      end

      @logger.info("Daemon started: #{@transfer_daemon_process}")

      return unless @server_port.zero?

      last_line = File.readlines(@daemon_log).last
      log_info = JSON.parse(last_line)
      port_match = log_info['msg'].match(/:(\d+)/)
      raise 'Could not read listening port from log file' unless port_match

      @server_port = port_match[1]
      @logger.info("Allocated server port: #{@server_port}")
    end

    def connect_to_daemon
      channel_address = "#{@server_address}:#{@server_port}"
      @logger.info("Connecting to #{@daemon_name} on: #{channel_address} ...")

      begin
        # GRPC::Core::Channel.new(channel_address, nil, :this_channel_is_insecure)
        @transfer_service = ::Transferd::Api::TransferService::Stub.new(channel_address, :this_channel_is_insecure)
        # Initiate actual connection
        get_info_response = @transfer_service.get_info(::Transferd::Api::InstanceInfoRequest.new)
        @logger.debug("Daemon info: #{get_info_response}")
      rescue GRPC::BadStatus => e
        @logger.error("Failed to connect: #{e}")
        raise 'failed to connect.'
      end

      @logger.info('Connected!')
    end

    def startup
      if @transfer_service.nil?
        start_daemon
        connect_to_daemon
      end
      self
    end

    def shutdown
      if @transfer_daemon_process
        @logger.info('Shutting down daemon...')
        Process.kill('KILL', @transfer_daemon_process)
        Process.wait(@transfer_daemon_process)
        @transfer_daemon_process = nil
      end
      @transfer_service = nil
    end

    def start_transfer(transfer_spec)
      ts_json = JSON.dump(transfer_spec)
      @logger.debug("ts: #{ts_json}")

      transfer_request = ::Transferd::Api::TransferRequest.new(
        transferType: ::Transferd::Api::TransferType::FILE_REGULAR,
        config: ::Transferd::Api::TransferConfig.new,
        transferSpec: ts_json
      )

      transfer_response = @transfer_service.start_transfer(transfer_request)
      throw_on_error(transfer_response.status, transfer_response.error)
      transfer_response.transferId
    end

    def wait_transfer(transfer_id)
      @logger.debug("transfer started with id #{transfer_id}")
      registration_request = ::Transferd::Api::RegistrationRequest.new(
        filters: [::Transferd::Api::RegistrationFilter.new(transferId: [transfer_id])]
      )
      @transfer_service.monitor_transfers(registration_request).each do |transfer_response|
        status = transfer_response.status
        # @logger.info("transfer: #{::Transferd::Api::TransferStatus.constants[status]}")
        @logger.info("transfer: #{status}")
        throw_on_error(status, transfer_response.transferInfo)
        break if status == :COMPLETED
      end
      @logger.info("Transfer #{transfer_id} completed successfully.")
    end

    def start_transfer_and_wait(t_spec)
      startup
      wait_transfer(start_transfer(t_spec))
    end

    def throw_on_error(status, info)
      if status == :FAILED
        raise "transfer failed: #{info.errorDescription}"
      elsif status == :UNKNOWN_STATUS
        raise "unknown transfer id: #{info.errorDescription}"
      end
    end

    private

    def ascp_level(level_string)
      case level_string
      when 'info'  then 0
      when 'debug' then 1
      when 'trace' then 2
      else
        raise "Invalid ascp_level: #{level_string}"
      end
    end
  end
end
